File: python/day10.py
from collections import deque
from functools import reduce
from operator import xor


def knot_logic(sizes, iterations=64):
    circle = deque(range(256))
    skip = 0
    for _ in range(iterations):
        for group_size in sizes:
            knot = [circle.popleft() for _ in range(group_size)]
            circle += reversed(knot)
            circle.rotate(-skip)
            skip += 1
    unwind = iterations * sum(sizes) + skip * (skip-1) // 2
    circle.rotate(unwind)
    return list(circle)


def knot_hashing(word):
    ascii_sizes = [ord(c) for c in word] + [17, 31, 73, 47, 23]
    numbers = knot_logic(ascii_sizes)
    SIZE = 256
    BLOCK_SIZE = 16
    block = lambda i: numbers[i*BLOCK_SIZE : (i+1)*BLOCK_SIZE]
    dense = [reduce(xor, block(i)) for i in range(SIZE // BLOCK_SIZE)]
    return ''.join(f'{n:02x}' for n in dense)


def main():
    with open('./inputs/10.txt') as f:
        input_string = f.readline()
    int_sizes = [int(n) for n in input_string.split(',')]

    first_hash = knot_logic(int_sizes, iterations=1)
    first = first_hash[0] * first_hash[1]
    print(first)

    second = knot_hashing(input_string)
    print(second)


if __name__ == '__main__':
    main()

File: nim/day10.nim
import sugar, strutils, sequtils

const res = lc[i | (i <- 0 .. 255), int]


func knotLogic(sizes: seq[int], iterations = 64): seq[int] =
  result = res
  const size = 255
  var pos, skip, a, b: int

  for i in 1 .. iterations:
    for groupSize in sizes:
      for p in 0 ..< (groupSize div 2):
        a = (pos + p) and size
        b = (pos + groupSize-1 - p) and size
        swap(result[a], result[b])
      pos = (pos + groupSize + skip) and size
      inc skip


func knotHashing*(word: string, binOut = false): string =
  let
    asciiSizes = word.mapIt(ord it).concat(@[17, 31, 73, 47, 23])
    numbers = knotLogic(asciiSizes)
  for bl in 0 .. 15:
    var hashed: int
    for i in 0 .. 15:
      hashed = hashed.xor(numbers[16*bl + i])

    let hashOut = if binOut: hashed.toBin(8) else: hashed.toHex(2)
    result.add(hashOut)


when isMainModule:
  let
    inputString = readFile("./inputs/10.txt")
    intSizes = inputString.split(',').map(parseInt)
    firstHash = knotLogic(intSizes, 1)
    first = firstHash[0] * firstHash[1]
    second = knotHashing(inputString).toLowerAscii()

  echo first
  echo second
